File: 621-task-scheduler/task-scheduler.cpp
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        std::unordered_map<char,int> freq;
        for (char elem : tasks) freq[elem]++;

        std::priority_queue<std::pair<int,char>> pq;
        for (const auto& [elem,count] : freq) pq.push({count,elem});

        std::queue<std::pair<int,std::pair<int,char>>> q;
        int time = 0;

        while (!pq.empty() || !q.empty()) {
            time++;

            // FIXED: Bring back all tasks ready at this time
            while (!q.empty() && q.front().first == time) {
                pq.push(q.front().second);
                q.pop();
            }

            if (!pq.empty()) {
                auto [count, elem] = pq.top();
                pq.pop();
                if (count - 1 > 0) {
                    q.push({time + n + 1, {count - 1, elem}});
                }
            }
        }

        return time;
    }
};

File: 621-task-scheduler/task-scheduler.py
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        freq = Counter(tasks)
        pq = [-val for val in freq.values()]
        heapq.heapify(pq)
        time = 0
        queue = deque()
        while pq or queue:
            time+=1
            if pq:
                count = heapq.heappop(pq)                
                if count+1 != 0:
                    queue.append([time+n,count+1])
            if queue and queue[0][0]==time:
                heapq.heappush(pq, queue.popleft()[1])            

        
        return time



        